README.md
# Lock for Composer

[![Continuous Integration](https://github.com/ghostwriter/composer-lock/actions/workflows/continuous-integration.yml/badge.svg)](https://github.com/ghostwriter/composer-lock/actions/workflows/continuous-integration.yml)
[![Supported PHP Version](https://badgen.net/packagist/php/ghostwriter/composer-lock?color=8892bf)](https://www.php.net/supported-versions)
[![Type Coverage](https://shepherd.dev/github/ghostwriter/composer-lock/coverage.svg)](https://shepherd.dev/github/ghostwriter/composer-lock)
[![Latest Version on Packagist](https://badgen.net/packagist/v/ghostwriter/composer-lock)](https://packagist.org/packages/ghostwriter/composer-lock)
[![Downloads](https://badgen.net/packagist/dt/ghostwriter/composer-lock?color=blue)](https://packagist.org/packages/ghostwriter/composer-lock)

Bumps `./composer.lock` file, optionally lock the minimum supported PHP version.

## Installation

You can install the package via composer:

``` bash
composer require ghostwriter/composer-lock
```

## Usage

``` bash
# Update lock file

composer lock

# Update lock file with 7.4 as the minimum supported PHP version

composer lock --php 7.4

# simulate updating the lock file with 8.0.999 as the minimum supported PHP version

composer lock -p 8.0.999 --dry-run
```

### Testing

``` bash
composer test
```

### Changelog

Please see [CHANGELOG.md](./CHANGELOG.md) for more information what has changed recently.

### Security

If you discover any security related issues, please email `nathanael.esayeas@protonmail.com` instead of using the issue tracker.

## Credits

- [Composer](https://github.com/composer/composer)
- [Nathanael Esayeas](https://github.com/ghostwriter)
- [All Contributors](../../contributors)

## License

The BSD-3-Clause. Please see [License File](./LICENSE) for more information.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file, in reverse chronological order by release.

## 0.3.2 - TBD

### Added

- Nothing.

### Changed

- Nothing.

### Deprecated

- Nothing.

### Removed

- Nothing.

### Fixed

- Nothing.

## 0.3.1 - 2022-08-13


-----

### Release Notes for [0.3.1](https://github.com/ghostwriter/composer-lock/milestone/14)

0.3.x bugfix release (patch)

### 0.3.1

- Total issues resolved: **0**
- Total pull requests resolved: **1**
- Total contributors: **1**

 - [20: Up](https://github.com/ghostwriter/composer-lock/pull/20) thanks to @ghostwriter

## 0.2.3 - 2022-08-12


-----

### Release Notes for [0.2.3](https://github.com/ghostwriter/composer-lock/milestone/12)

0.2.x bugfix release (patch)

### 0.2.3

- Total issues resolved: **0**
- Total pull requests resolved: **1**
- Total contributors: **1**

 - [18: Delete continuous-integration.yml](https://github.com/ghostwriter/composer-lock/pull/18) thanks to @ghostwriter

## 0.2.2 - 2022-08-12

-----

### Release Notes for [0.2.2](https://github.com/ghostwriter/composer-lock/milestone/10)

0.2.x bugfix release (patch)

### 0.2.2

- Total issues resolved: **0**
- Total pull requests resolved: **1**
- Total contributors: **1**

 - [16: Use GitHub PAT](https://github.com/ghostwriter/composer-lock/pull/16) thanks to @ghostwriter

## 0.1.0 - TBD


tests/Unit/LockTest.php
<?php

declare(strict_types=1);

namespace Ghostwriter\ComposerLock\Tests\Unit;

use Composer\Console\Application;
use Composer\Plugin\Capability\CommandProvider;
use Composer\Plugin\Capable;
use Composer\Plugin\PluginInterface;
use Generator;
use Ghostwriter\ComposerLock\Lock;
use PHPUnit\Framework\TestCase;
use Symfony\Component\Console\Command\Command;
use Symfony\Component\Console\Tester\CommandTester;

use function class_exists;
use function sprintf;
use function substr;
use function trim;

/**
 * @coversDefaultClass Lock
 */
final class LockTest extends TestCase
{
    private CommandTester $commandTester;
    private Lock $lock;
    protected function setUp(): void
    {
        $this->lock = new Lock();

        $this->commandTester = new CommandTester(
            (static function (Application $application, Lock $lock): Command {
                $application->setAutoExit(false);
                $application->add($lock->getCommands()[0]);

                return $application->find(Lock::PACKAGE_COMMAND);
            })(
                new Application(),
                $this->lock
            )
        );
    }

    public function testPluginExists(): void
    {
        self::assertTrue(class_exists(Lock::class));
    }

    public function testImplementsPluginInterface(): void
    {
        self::assertInstanceOf(PluginInterface::class, $this->lock);
    }

    public function testImplementsCommandProviderInterface(): void
    {
        self::assertInstanceOf(CommandProvider::class, $this->lock);
    }

    public function testImplementsCapableInterface(): void
    {
        self::assertInstanceOf(Capable::class, $this->lock);
    }

    /**
     * @dataProvider supportedCommandsProvider
     * @param array<"--dry-run"|"--php"|"-d"|"-p"|"command", bool|string> $input
     * @param array<"capture_stderr_separately"|"verbosity", bool> $options
     */
    public function testCommandOptionsProvider(
        array $input,
        array $options,
        string $expectedOutput = '',
        string $expectedErrorOutput = '',
        int $expectedStatusCode = 0
    ): void {
        $this->commandTester->execute($input, $options);

        self::assertSame($this->commandTester->getDisplay(true), $expectedOutput);
        self::assertSame($this->commandTester->getErrorOutput(true), $expectedErrorOutput);
//        self::assertSame($this->commandTester->getStatusCode(), $expectedStatusCode);
    }

    /**
     * @doesNotPerformAssertions
     * phpcs:disable
     * @return Generator<string, array{0: array<"--dry-run"|"--php"|"-d"|"-p"|"command", bool|string>, 1: array{capture_stderr_separately: true, verbosity: true}, 2: '', 3: '', 4: 0}, mixed, void>
     * phpcs:enable
     */
    public function supportedCommandsProvider(): Generator
    {
        foreach ($this->phpCommandOptions() as $phpCommandOption) {
            foreach ($this->supportedPhpVersions() as $phpVersion) {
                $phpVersionKey = substr($phpVersion, 0, 3);
                foreach ($this->packageFixtures() as $version => $packageFixture) {
                    if ($version !== $phpVersionKey) {
                        continue;
                    }
                    foreach ($this->dryRunCommandOptions() as $dryRun) {
                        yield  trim(sprintf(
                            'composer lock %s %s -d %s %s',
                            $phpCommandOption,
                            $phpVersion,
                            $packageFixture,
                            $dryRun ? '--dry-run' : ''
                        )) => [
                            [
                                'command'         => 'lock',
                                $phpCommandOption => $phpVersion,
                                '-d'              => $packageFixture,
                                '--dry-run'       => $dryRun,
                            ],
                            ['capture_stderr_separately' => true, 'verbosity' => true],
                            '',
                            '',
                            0,
                        ];
                    }
                }
            }
        }
    }

    /**
     * @return Generator<int, bool, mixed, void>
     */
    private function dryRunCommandOptions(): Generator
    {
        yield from [true, false];
    }

    /**
     * @return Generator<int, '--php'|'-p', mixed, void>
     */
    private function phpCommandOptions(): Generator
    {
        yield from ['--php', '-p'];
    }

    /**
     * @return Generator<int, string, mixed, void>
     */
    private function supportedPhpVersions(): Generator
    {
        yield from ['7.4', '8.0', '8.1', '7.4.999', '8.0.999', '8.1.999'];
    }

    /**
     * @return Generator<'7.4'|'8.0'|'8.1', string, mixed, void>
     */
    private function packageFixtures(): Generator
    {
        yield from [
            '7.4' => __DIR__ . '/../fixtures/package-74',
            '8.0' => __DIR__ . '/../fixtures/package-80',
            '8.1' => __DIR__ . '/../fixtures/package-81',
        ];
    }
}


src/Lock.php
<?php

declare(strict_types=1);

namespace Ghostwriter\ComposerLock;

use Composer\Command\BaseCommand;
use Composer\Composer;
use Composer\Console\Application;
use Composer\InstalledVersions;
use Composer\IO\IOInterface;
use Composer\Plugin\Capability\CommandProvider;
use Composer\Plugin\Capable;
use Composer\Plugin\PluginInterface;
use Symfony\Component\Console\Input\ArrayInput;
use Symfony\Component\Console\Input\InputInterface;
use Symfony\Component\Console\Input\InputOption;
use Symfony\Component\Console\Output\OutputInterface;
use Throwable;

use function implode;
use function sprintf;
use function substr;

use const PHP_EOL;

final class Lock implements PluginInterface, Capable, CommandProvider
{
    /**
     * @var string
     */
    public const PACKAGE_NAME = 'ghostwriter/composer-lock';

    /**
     * @var string
     */
    public const PACKAGE_AUTHOR = 'Nathanael Esayeas';

    /**
     * @var string
     */
    public const PACKAGE_COMMAND = 'lock';

    /**
     * @var string
     */
    public const PACKAGE_ATTRIBUTION = '<info>%s</info> (<comment>%s@%s</comment>) by <info>%s</info> and <comment>contributors</comment>.'; //phpcs:ignore

    /**
     * @var string
     */
    public const COMMAND_SUCCESS = '[<comment>%s</comment>]%s';

    /**
     * @var string
     */
    public const COMMAND_FAILURE = '[<comment>%s</comment>]<error>%s</error>';

    /**
     * @var int
     */
    public const SHORT_SHA_LENGTH = 7;

    private BaseCommand $lockCommand;

    public function __construct()
    {
        $this->lockCommand = new class extends BaseCommand
        {
            public function __construct()
            {
                parent::__construct(Lock::PACKAGE_COMMAND);
            }

            protected function configure(): void
            {
                $this->setName(Lock::PACKAGE_COMMAND)
                    ->setDescription('Manage composer.lock file for any PHP version.')
                    ->setHelp(
                        implode(PHP_EOL, [
                            'Use "--php|-p" to set the minimum supported PHP version.',
                            'e.g. "7.2", "7.3", "7.4", "8.0", "8.1", "8.2"',
                        ])
                    )
                    ->setDefinition([
                        new InputOption(
                            'php',
                            'p',
                            InputOption::VALUE_REQUIRED,
                            'Set the minimum supported PHP version.',
                            null
                        ),
                        new InputOption(
                            'dry-run',
                            null,
                            InputOption::VALUE_NONE,
                            'Display outputs without execute anything (implicitly enables --verbose).',
                        ),
                    ]);
            }

            public function error(string $message): void
            {
                $this->getIO()->writeError(sprintf(
                    Lock::COMMAND_FAILURE,
                    Lock::PACKAGE_NAME,
                    $message
                ));
            }

            public function success(string $message): void
            {
                $this->getIO()->write(sprintf(
                    Lock::COMMAND_SUCCESS,
                    Lock::PACKAGE_NAME,
                    $message
                ));
            }

            public function write(string $message): void
            {
                $this->getIO()->write($message);
            }

            /** @psalm-return 0|1 */
            protected function execute(InputInterface $input, OutputInterface $output): int
            {
                $packageVersion   = (string) InstalledVersions::getPrettyVersion(Lock::PACKAGE_NAME);
                $packageReference = substr(
                    (string) InstalledVersions::getReference(Lock::PACKAGE_NAME),
                    0,
                    Lock::SHORT_SHA_LENGTH
                );

                $this->write(sprintf(
                    'Running ' . Lock::PACKAGE_ATTRIBUTION,
                    Lock::PACKAGE_NAME,
                    $packageVersion,
                    $packageReference,
                    Lock::PACKAGE_AUTHOR
                ));

                /** @var string|null $phpVersion */
                $phpVersion = $input->getOption('php');

                /** @var bool $dryRun */
                $dryRun = $input->getOption('dry-run');

                try {
                    $this->configCommand(
                        'Updating "config.platform.php" in composer.json',
                        'Failed to update "config.platform.php" in composer.json file',
                        new ArrayInput([
                            'setting-key'   => 'platform.php',
                            'setting-value' => [$phpVersion],
                        ]),
                        $output,
                        $dryRun,
                        $phpVersion
                    );

                    $this->updateCommand(
                        'Updating composer dependencies.',
                        'Failed to update composer dependencies',
                        new ArrayInput([
                            '--no-autoloader' => true,
                            '--no-cache'      => true,
                            '--no-scripts'    => true,
                            '--no-plugins'    => true,
                            '--no-progress'   => true,
                            '--quiet'         => true,
                        ]),
                        $output,
                        $dryRun
                    );

                    $this->configCommand(
                        'Removing "config.platform.php" in composer.json',
                        'Failed to remove "config.platform.php" in composer.json',
                        new ArrayInput([
                            '--unset'     => true,
                            'setting-key' => 'platform.php',
                        ]),
                        $output,
                        $dryRun,
                        $phpVersion
                    );

                    $this->updateCommand(
                        'Updating composer.lock',
                        'Failed to update composer.lock',
                        new ArrayInput([
                            '--no-cache'    => true,
                            '--no-scripts'  => true,
                            '--no-plugins'  => true,
                            '--no-progress' => true,
                            '--quiet'       => true,
                            '--no-install'  => true,
                            '--lock'        => true,
                        ]),
                        $output,
                        $dryRun
                    );
                } catch (Throwable $throwable) {
                    return 1;
                }

                $this->success('Successfully updated composer.lock');

                return 0;
            }

            private function configCommand(
                string $success,
                string $error,
                InputInterface $input,
                OutputInterface $output,
                bool $dryRun = false,
                ?string $phpVersion = null
            ): void {
                if ($phpVersion === null) {
                    return;
                }

                $this->success($success);

                if ($dryRun) {
                    return;
                }

                try {
                    (function (
                        Application $application,
                        InputInterface $input,
                        OutputInterface $output
                    ): void {
                        $application->resetComposer();
                        $application->find('config')->run($input, $output);
                    })($this->getApplication(), $input, $output);
                } catch (Throwable $throwable) {
                    $this->error($error . ', ' . $throwable->getMessage());

                    throw $throwable;
                }
            }

            private function updateCommand(
                string $success,
                string $error,
                InputInterface $input,
                OutputInterface $output,
                bool $dryRun = false
            ): void {
                try {
                    $this->success($success);

                    if ($dryRun) {
                        return;
                    }

                    (function (
                        Application $application,
                        InputInterface $input,
                        OutputInterface $output
                    ): void {
                        $application->resetComposer();
                        $application->find('update')->run($input, $output);
                    })($this->getApplication(), $input, $output);
                } catch (Throwable $throwable) {
                    $this->error($error . ', ' . $throwable->getMessage());

                    throw $throwable;
                }
            }
        };
    }

    public function activate(Composer $composer, IOInterface $io): void
    {
    }

    public function deactivate(Composer $composer, IOInterface $io): void
    {
    }

    /**
     * @return array{CommandProvider::class: self::class}
     */
    public function getCapabilities(): array
    {
        return [CommandProvider::class => self::class];
    }

    /** @return array{0: BaseCommand} */
    public function getCommands(): array
    {
        return [$this->lockCommand];
    }

    public function uninstall(Composer $composer, IOInterface $io): void
    {
    }
}
